File: challenges/Ferrum 2018/solution2.py
def solution(A):
    n = len(A)
    # 前缀和数组
    p = [0] * (n + 1)
    # 以这一项结尾的子数组的最大和
    best = [0] * n
    length, m = 0, 0
    for i in range(1, n + 1):
        p[i] = p[i - 1] + A[i - 1]
        best[i - 1] = p[i] - p[m]
        while length < i and best[i - length - 1] + p[i] - p[i - length] >= 0:
             length += 1
        if p[i] < p[m]:
            m = i
    return length

File: challenges/Ferrum 2018/Solution2.java
// you can also use imports, for example:
// import java.util.*;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");

class Solution {
    public int solution(int[] A) {
        // Implement your solution here
        int n = A.length;
        // 前缀和数组
        int[] p = new int[n + 1];
        // 以这一项结尾的子数组的最大和
        int[] best = new int[n];
        int length = 0;
        for (int i = 1, m = 0; i <= n; ++i) {
            p[i] = p[i - 1] + A[i - 1];
            best[i - 1] = p[i] - p[m];
            for (; length < i && best[i - length - 1] + p[i] - p[i - length] >= 0; ++length)
            ;
            if (p[i] < p[m]) {
                m = i;
            }
        }
        return length;
    }
}
